CLN: remove util._decorators.make_signature and make related changes

topper-123 opened the pull request:
``util._decorators.make_signature`` is only used in ``core.groupby.base.whitelist_method_generator``.

However, in that function it is only used inside an if statement, that is conditional upon an attribute being a method. That however can not happen, as the function parameter only takes classes as an input, so that if-statement is never true and its content never executed. The net effect is that a nice chunk of code can just be removed, which is nice.

In the future, we should just use jus use ``inspect.signature`` or ``inspect.getfullargspecs`` for signature inspection.

topper-123 committed:
CLN: remove util._decorators.make_signature and make related changes

topper-123 commented:
The failures are unrelated to this PR

topper-123 commented on pandas/tests/util/test_util.py:
```diff
@@ -37,22 +35,6 @@ def test_numpy_err_state_is_default():
-                       ["old_arg_name", "new_arg_name",
-                        "mapping", "stacklevel"]))
-])
-def test_make_signature(func, expected):
```
It can be noted that ``validate_kwargs`` and ``deprecate_kwarg`` have their own test modules, so removing here it not an issue.

WillAyd commented:
Makes sense to me. I think merging master should get rid of the failure.

Also annotations while you are looking at this would be great if they are easy to add

topper-123 committed:
Add typing information

topper-123 commented:
I've rebased and added type hints.

Wrt. the type hints, I've added the typing specific internal imports inside a ``if TYPE_CHECKING`` statement in order to minimize the risk of circular import problems. I'm however not 100 % strong in correct typing usage, so I see that as point for discussion, not ny final view on pure type importing.

WillAyd commented on pandas/core/groupby/base.py:
```diff
@@ -3,13 +3,14 @@
 
 from pandas.core.dtypes.common import is_list_like, is_scalar
 
+if TYPE_CHECKING:
```
Was this actually causing circular imports? We import DataFrame and Series elsewhere within this directory didn't think those would cause issues.

Would like to avoid the TYPE_CHECKING variable generally

WillAyd replied:
Note this probably won't matter in light of other comments below

topper-123 replied:
No, it wasn't. Just my nervousness that it might cause one in a future commit.

topper-123 replied:
I don't understand you here. Could you be more explicit?

jreback replied:
where a type is expected, for example DataFrame, you can simply use the string 'DataFrame' rather than importing it (this only works for types :->) e.g.

```
def func(df: 'DataFrame) -> 'DataFrame':
   ...
```

is legit. Of course we prefer to import, but we need to avoid circular deps (e.g. in groupby in particular).

topper-123 replied:
Yes, it's legit, but it won't be infer correctly without the import?

At least in PyCharm it needs an import to recognize the type hint types do proper code completions, even when quoted.

topper-123 replied:
I've checked again, and actually if I remove the ``if TYPE_CHECKING`` statement, it will cause a circular import error.

So I think the if-statement must remain in order to do type checking. Also, the DataFrame rather than ABCDataFrame gives better results for PyCharm, without negativ results for mypy, so I propose keeping it that way.

Although adding type hints does feel like walking in uncharted territory as far as figuring out what best-practice is.

WillAyd replied:
Yea I think this will be required for "GroupBy" since it is not in the same module. FWIW you can do this by putting `reveal_type(base)` in the body of `whitelist_method_generator`.

Doing `Type['GroupBy']` yields the below errors:

```sh
$ mypy pandas/core/groupby/base.py
pandas/core/groupby/base.py:93: error: Name 'GroupBy' is not defined
pandas/core/groupby/base.py:124: error: Revealed type is 'Type[Any]'
```

Whereas this import for mypy infers correctly:

```sh
$ mypy pandas/core/groupby/base.py
pandas/core/groupby/base.py:127: error: Revealed type is 'Type[pandas.core.groupby.groupby.GroupBy]'
```

So I think this is correct for GroupBy (separate conversations still for DataFrame and Series)

WillAyd commented on pandas/core/groupby/base.py:
```diff
@@ -93,19 +94,21 @@ def _gotitem(self, key, ndim, subset=None):
 
 
-def whitelist_method_generator(base, klass, whitelist):
+def whitelist_method_generator(base: 'Type[GroupBy]',
```
While this is probably just used for GroupBy probably OK to just type this as `Type`, i.e. it could theoretically accept any Type not just a GroupBy

jreback replied:
this is pretty descriptive here, if the more specific types work that is good

WillAyd replied:
OK sure. Though I think should be `Type['GroupBy']` then (note difference in quoting)

topper-123 replied:
My preference is an all-or-nothing approach to quoting. Else it's more work to check if a annotation is quoted or not.

WillAyd replied:
Not sure I totally follow your point but the intention of quoting is to provide forward references to objects perhaps before they are defined. There's no reason to do this with building typing objects

Here's the mypy documentation on that. You'll see a particular example where they populate a List with a forward reference, where the quoting only apply to the type variable. That's what I'm asking for here:

https://mypy.readthedocs.io/en/latest/kinds_of_types.html#class-name-forward-references

WillAyd replied:
Though may not even want this at all:

> String literal types must be defined (or imported) later in the same module. They cannot be used to leave cross-module references unresolved. (For dealing with import cycles, see Import cycles.)

Don't need this to be a blocker

WillAyd replied:
Another option here if we don't like the import machinery is to just move this to generic where it is actually used. I don't see a very strong reason why this needs to be in base

topper-123 replied:
I've moved ``whitelist_method_generator`` to generic.py. That does make the typing simpler. I've also added the TypeVar in _typing.

However, I suggest that unless we're fixing relatively simple typing issues (no new defined types in ``_typing.py``  etc.), that typing changes go in separate commits?

jreback replied:
separate commits are certainly fine for typing but not a PR unless major work needs to be done; we get much more mileage out of adding typing as we going in changed code (granted sometimes this will be slightly painful at the beginning)

WillAyd commented on pandas/core/groupby/base.py:
```diff
@@ -93,19 +94,21 @@ def _gotitem(self, key, ndim, subset=None):
 
-def whitelist_method_generator(base, klass, whitelist):
+def whitelist_method_generator(base: 'Type[GroupBy]',
+                               klass: 'Union[Type[DataFrame], Type[Series]]',
```
This should probably be a TypeVar instead of a Union (i.e. we parametrize on one or the other of these, but don't use interchangeably within the function).

This is going to be a common idiom - could you add `FrameOrSeries = TypeVar('FrameOrSeries', ABCSeries, ABCDataFrame)` to pandas._typing and use that here instead (open to other naming suggestions)

topper-123 replied:
I don't understand TypeVars all too well, but my understanding is it it ensures that the same type comes out as came in. So, if a Series came in, and an object of the same TypeVar comes out, that means the returnm type is a Series (and not DataFrame). In that case, it will not matter in this case if we use Union or TypeVar, isn't that true?

Not opposed to using TypeVar, just trying to understand their use cases.

WillAyd replied:
Yea you've got that right! This comment might also be helpful

https://github.com/pandas-dev/pandas/issues/26453#issuecomment-497025873

So I don't think will make a difference here - just looking to get some kind of consistency in place as our annotations are still relatively immature

topper-123 replied:
Ok. It will be actually be more concise to use that TypeVar here (fewer type-specific imports), and a ``TypeVar('FrameOrSeries', Series, DataFrame)``  would make sense for Pandas in general.

In Pycharm ``TypeVar('FrameOrSeries', ABCSeries, ABCDataFrame)`` doesn't give the correct completions, while ``TypeVar('FrameOrSeries', Series, DataFrame)`` does give the right ones. any specific reason to use ABCSeries and ABCDataFrame over the actual classes?

I'm also thinking that the best version would be actually  ``TypeVar('FrameOrSeries', NDFRame, Series, DataFrame)`, just to be able to add type relevant hints to the NDFrame class. Any thoughts on this?

WillAyd replied:
ABCSeries and ABCDataFrame are supposed to circumvent issues with import cycles. Unfortunately I don't use PyCharm so I can't speak to their support for those w.r.t. code completion.

I also don't think NDFrame should be a part of that as it's not a type that is used directly (i.e. typically only used as a superclass)

WillAyd replied:
> ABCSeries and ABCDataFrame are supposed to circumvent issues with import cycles.

To clarify though, if you see a better way of doing this certainly open to suggestions. I can't claim total expertise in our comprehensive usage of the ABC classes. @jreback

topper-123 replied:
@WillAyd, do you use VSCode? Do you get correct type hints from TypeVars?

WillAyd replied:
No. Haven't used annotations for code completion so any insights you can offer there would be appreciated

topper-123 commented:
Given that a Union works as well as a TypeVar in this case, and the required TypeVar would actually be ``TypeVar('FrameOrSeries', Type[ABCSeries], Type[ABCDataFrame])`` (notice the extra ``Type``), and this TypeVar would be used very infrequently, I propose just accepting it like is it ATM (a Union).

WillAyd commented:
Might be mixing up a few things here. The TypeVar would look as follows:

```python
FrameOrSeries = TypeVar('FrameOrSeries', ABCSeries, ABCDataFrame)
```

The annotation in question would then be `Type[FrameOrSeries]` as opposed to `Type[Union[Series, DataFrame]]]` as is now (perhaps some slight variations).

topper-123 committed:
moved base.whitelist_method_generator to .generic

WillAyd commented:
Change looks good not sure what's going on with Travis / NumPy failure

jreback commented on pandas/_typing.py:
```diff
@@ -22,3 +22,5 @@
                              Timedelta)
 Dtype = Union[str, np.dtype, ExtensionDtype]
 FilePathOrBuffer = Union[str, Path, IO[AnyStr]]
```
@WillAyd we should put a line comment on when to use these / what these are if not completely obvious (and then maybe even so).

future PR of course

WillAyd replied:
Yea I'm going to take another look at this module soon to see if there's a better way of doing things. Also wondering if we should use `if TYPE_CHECKING` imports in that module instead of the ABCs. Will check it out and post separately

jreback replied:
> Yea I'm going to take another look at this module soon to see if there's a better way of doing things. Also wondering if we should use `if TYPE_CHECKING` imports in that module instead of the ABCs. Will check it out and post separately

-1 on this. I think the way you have it setup is very nice actually, unless I am missing something big.

WillAyd replied:
My concern is that the ABC* classes aren't getting type checked correctly. Adding `reveal_type(klass)` to this PR where the TypeVar is used yields the following:

```sh
$ mypy pandas/core/groupby/generic.py
pandas/core/groupby/generic.py:83: error: Revealed type is 'Type[Any]'
```

That coupled with @topper-123 's comment around code completion makes me think the ABCs might all be `Type[Any]` since they are generated by `create_pandas_abc_type`, a function without any annotations itself (so implicitly `Type[Any]`)

Going to investigate and open a separate issue accordingly

jreback replied:
ok fair. I just don't want to introduce overhead for contributors, which typing especially with 'non-standard' things (e.g. ``cast``) can do; we really want to minimize what people have to do (nothwithstanding maybe we have magic that makes things easy).

jreback commented:
lgtm. I restarted the job, should pass. merge on green.

jreback commented:
thanks @topper-123
